### DISCUSSION TITLE
Lock leading to race condition with a simple SELECT query while using QueuePool & fastapi_async_sqlalchemy

### DISCUSSION THREAD

**Discussion by neeagl** [sqlalchemy/sqlalchemy#10890] [Usage Questions] [answered]
Hi,
I'm trying to make a simple query:
```
db_session = get_db()
users = await db_session.execute(select(User).where(User.email == email))
```

This is how the DB is initialized:

```
engine = create_async_engine(
    str(settings.ASYNC_DATABASE_URI),
    # echo=True,
    future=True,
    pool_pre_ping=True,
    pool_size=POOL_SIZE,
    max_overflow=64,
    poolclass=NullPool
    if settings.MODE == ModeEnum.testing # TODO: need to make it NullPool for celery related jobs as well
    else QueuePool,  # Asincio pytest works with NullPool
)

SessionLocal = sessionmaker(
    autocommit=False,
    autoflush=False,
    bind=engine,
    class_=AsyncSession,
    expire_on_commit=False,
)

async def get_db() -> AsyncGenerator[AsyncSession, None]:
    async with SessionLocal() as session:
        yield session
```

FastAPI middleware:
```

app.add_middleware(
    SQLAlchemyMiddleware,
    db_url=str(settings.ASYNC_DATABASE_URI),
    engine_args={
        # "echo": True,
        "pool_pre_ping": True,
        "pool_size": settings.POOL_SIZE,
        "max_overflow": 64,
        "poolclass": NullPool
        if settings.MODE == ModeEnum.testing
        else QueuePool,  # Asincio pytest works with NullPool
    },
)

```

When I'm testing it with a set of 8-10 concurrent API hits, it gets into a race condition due to a lock.

**Comment by CaselIT** [comment]
Hi,

QueuePool is not the correct class to use with asyncio.
Your question it too vague to provide any other help

**Reply by neeagl** [reply]
I have updated the question with more info. Kindly let me know your thought.

**Comment by zzzeek** [comment]
dont send "poolclass" if you want the default queuepool to be used, `create_async_engine()` will use the correct pool automatically

**Reply by CaselIT** [reply]
at some point I think that a check should be added for this

**Reply by neeagl** [reply]
I have updated the question with more info. Kindly let me know your thought.

**Reply by zzzeek** [reply, marked as answer]
no change, dont do this:

```
 poolclass=NullPool
    if settings.MODE == ModeEnum.testing # TODO: need to make it NullPool for celery related jobs as well
    else QueuePool,  # Asincio pytest works with NullPool

```

that is wrong, do not use QueuePool with asyncio

**Reply by neeagl** [reply]
Great that works, thank you so much. Is there any documentation that I can read to understand what was happening here? Just starting out with async.

**Reply by zzzeek** [reply]
basic docs are at https://docs.sqlalchemy.org/en/20/orm/extensions/asyncio.html#synopsis-core   ...basically you are using a pooling API that is not really mentioned in any of our examples that I'm familiar with so maybe that's from a blog post or something?
